### PR TITLE
Displaying Warning when source(s) doesn't match with the Test Run platform/framework.

### PR DESCRIPTION
**Pull request by navin22**
-  Comparing the inferred platform & framework  with default platform and framework chosen and displaying warning when differs.
- Added Unit test, Acceptance tests and fixed some existing tests.

### CODE REVIEW

**Comment by navin22** [comment]
@dotnet-bot test Windows_NT / Debug Build

File: src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs [hunk 1]
```diff
@@ -410,5 +411,88 @@ public static bool TryGetFrameworkXml(XPathNavigator runSettingsNavigator, out s
...
+        /// Returns the sources matching the specified platform and framework settings.
+        /// For incompatible sources, warning is added to incompatibleSettingWarning.
+        /// </summary>
+        public static IEnumerable<String> FilterCompatibleSources(Architecture chosenPlatform,
+            Framework chosenFramework,
+            IDictionary<String,Architecture> sourcePlatforms,
+            IDictionary<String, Framework> sourceFrameworks,
+            out String incompatibleSettingWarning)
+        {
+            incompatibleSettingWarning = null;
```
Comment (codito) [inline src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs]: nit: suggest string.Empty.

File: src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs [hunk 2]
```diff
@@ -410,5 +411,88 @@ public static bool TryGetFrameworkXml(XPathNavigator runSettingsNavigator, out s
...
+            List<String> compatibleSources = new List<String>();
+            StringBuilder warnings = new StringBuilder();
+            warnings.AppendLine("");
```
Comment (codito) [inline src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs]: nit: just `AppendLine()` without args?

File: src/vstest.console/TestPlatformHelpers/TestRequestManager.cs [hunk 3]
```diff
@@ -354,18 +355,45 @@ private bool UpdateRunSettingsIfRequired(string runsettingsXml,List<string> sour
...
+                        inferedPlatform.ToString(), overwrite: true);
+                    }
+                    else
+                    {
+                        chosenPlatform = commandLineOptions.TargetArchitecture;
+                    }
+
+                    string incompatiableSettingWarning = string.Empty;
+
+                    var compatiableSources = InferRunSettingsHelper.FilterCompatibleSources(chosenPlatform, chosenFramework, sourcePlatforms, sourceFrameworks, out incompatiableSettingWarning);
```
Comment (codito) [inline src/vstest.console/TestPlatformHelpers/TestRequestManager.cs]: nit: compatible spelling

File: src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs [hunk 4]
```diff
@@ -410,5 +411,88 @@ public static bool TryGetFrameworkXml(XPathNavigator runSettingsNavigator, out s
...
+            }
+            return !sourceFramework.Equals(targetFramework);
+        }
+
+
+        /// <summary>
+        /// Returns the sources matching the specified platform and framework settings.
+        /// For incompatible sources, warning is added to incompatibleSettingWarning.
+        /// </summary>
+        public static IEnumerable<String> FilterCompatibleSources(Architecture chosenPlatform,
```
Comment (codito) [inline src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs]: nit: move public method above privates

File: src/Microsoft.TestPlatform.Utilities/Resources/Resources.resx [hunk 5]
```diff
@@ -117,6 +117,10 @@
   <resheader name="writer">
     <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
   </resheader>
+  <data name="DisplayChosenSettings" xml:space="preserve">
```
Comment (codito) [inline src/Microsoft.TestPlatform.Utilities/Resources/Resources.resx]: The actual text and translated text are already available in OM resources. Possible to reuse?

File: src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs [hunk 6]
```diff
@@ -410,5 +411,88 @@ public static bool TryGetFrameworkXml(XPathNavigator runSettingsNavigator, out s
...
+            Framework chosenFramework,
+            IDictionary<String,Architecture> sourcePlatforms,
```
Comment (codito) [inline src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs]: nit: space after comma

File: src/vstest.console/TestPlatformHelpers/TestRequestManager.cs [hunk 7]
```diff
@@ -354,18 +355,45 @@ private bool UpdateRunSettingsIfRequired(string runsettingsXml,List<string> sour
...
+                    }
 
+                    var inferedPlatform = inferHelper.AutoDetectArchitecture(sources, sourcePlatforms);
+                    var chosenPlatform = inferedPlatform;
                     if (updatePlatform)
                     {
                         InferRunSettingsHelper.UpdateTargetPlatform(navigator,
-                            inferHelper.AutoDetectArchitecture(sources).ToString(), overwrite: true);
-                        settingsUpdated = true;
+                        inferedPlatform.ToString(), overwrite: true);
```
Comment (codito) [inline src/vstest.console/TestPlatformHelpers/TestRequestManager.cs]: nit: indent, merge with above line

**Review by codito — APPROVED** [review]

File: src/Microsoft.TestPlatform.Utilities/Resources/Resources.resx (hunk 5 above)
Comment (navin22) [reply src/Microsoft.TestPlatform.Utilities/Resources/Resources.resx]: I didn't want to introduce the dependency ? Can i Delete it from there, as there's no reference to them.?

Comment (navin22) [reply src/Microsoft.TestPlatform.Utilities/Resources/Resources.resx]: Already there's dependency, so will remove it from UtilitiesResources.

File: src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs (hunk 4 above)
Comment (navin22) [reply src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs]: Done.

File: src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs (hunk 6 above)
Comment (navin22) [reply src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs]: Done

File: src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs (hunk 1 above)
Comment (navin22) [reply src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs]: DOne

File: src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs (hunk 2 above)
Comment (navin22) [reply src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs]: Done

File: src/vstest.console/TestPlatformHelpers/TestRequestManager.cs (hunk 7 above)
Comment (navin22) [reply src/vstest.console/TestPlatformHelpers/TestRequestManager.cs]: Done

File: src/vstest.console/TestPlatformHelpers/TestRequestManager.cs (hunk 3 above)
Comment (navin22) [reply src/vstest.console/TestPlatformHelpers/TestRequestManager.cs]: Done

File: src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs [hunk 8]
```diff
@@ -410,5 +412,83 @@ public static bool TryGetFrameworkXml(XPathNavigator runSettingsNavigator, out s
...
+
+            return sourcePlatform != targetPlatform;
+        }
+
+        /// <summary>
+        /// Returns true if source FrameworkVersion is incompatible with target FrameworkVersion.
+        /// </summary>
+        private static bool IsFrameworkIncompatible(Framework sourceFramework, Framework targetFramework)
+        {
+            if (sourceFramework.Version.Equals(Framework.DefaultFramework.Version))
```
Comment (smadala) [inline src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs]: Handle case sourceFramework is net45 and targetFramework is net451.

File: src/vstest.console/CommandLine/InferHelper.cs [hunk 9]
```diff
@@ -62,10 +63,8 @@ public Architecture AutoDetectArchitecture(List<string> sources)
 
                         if (!finalArch.Equals(arch))
                         {
-                            // TODO Log to console and client.
                             finalArch = Constants.DefaultPlatform;
                             EqtTrace.Info("Conflict in platform architecture, using default platform:{0}", finalArch);
-                            break;
                         }
```
Comment (smadala) [inline src/vstest.console/CommandLine/InferHelper.cs]: Log source and platform to EqtTrace.

File: src/vstest.console/TestPlatformHelpers/TestRequestManager.cs [hunk 10]
```diff
@@ -354,18 +355,43 @@ private bool UpdateRunSettingsIfRequired(string runsettingsXml,List<string> sour
...
+                    }
+
+                    string incompatiableSettingWarning = string.Empty;
+
+                    var compatibleSources = InferRunSettingsHelper.FilterCompatibleSources(chosenPlatform, chosenFramework, sourcePlatforms, sourceFrameworks, out incompatiableSettingWarning);
+
+                    if(!string.IsNullOrEmpty(incompatiableSettingWarning))
+                    {
+                        EqtTrace.Info(incompatiableSettingWarning);
+                        LoggerUtilities.RaiseTestRunWarning(this.testLoggerManager, this.testRunResultAggregator, incompatiableSettingWarning);
```
Comment (smadala) [inline src/vstest.console/TestPlatformHelpers/TestRequestManager.cs]: Will this print message in designmode too?

File: src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs [hunk 11]
```diff
@@ -36,6 +36,7 @@ public class InferRunSettingsHelper
         private const string TargetFrameworkNodePath = @"/RunSettings/RunConfiguration/TargetFrameworkVersion";
         private const string ResultsDirectoryNodePath = @"/RunSettings/RunConfiguration/ResultsDirectory";
         private const string TargetDeviceNodePath = @"/RunSettings/RunConfiguration/TargetDevice";
+        private const string multiTargettingForwardLink = @"http://go.microsoft.com/fwlink/?LinkID=236877&clcid=0x409";
```
[the remark this answers is not in the document]
Comment (navin22) [reply src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs]: Will ask @pvlakshm to point our forward link to this.

File: src/vstest.console/CommandLine/InferHelper.cs (hunk 9 above)
Comment (navin22) [reply src/vstest.console/CommandLine/InferHelper.cs]: Already logging the warning as is.

File: src/vstest.console/TestPlatformHelpers/TestRequestManager.cs (hunk 10 above)
Comment (navin22) [reply src/vstest.console/TestPlatformHelpers/TestRequestManager.cs]: Yes.

**Pull request merged**